mongo/collection.go

```diff
@@ -1503,7 +1436,14 @@ func (coll *Collection) Find(ctx context.Context, filter interface{},
 		rc = nil
 	}
 
-	fo := mergeFindOptions(opts...)
+	fo := &options.FindArgs{}
+	for _, opt := range opts {
+		for _, optFn := range opt.Opts {
+			if err := optFn(fo); err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
CC: @matthewdale 

We can avoid these type-specific merge requirements using generics. I propose that we add a new type to the `options` package that unify option-bearing structs:

```go
// Getter is an interface that wraps the GetOptions method to return a
// list of setters that can set data on the functional parameter of type T.
type Getter[T Optioners] interface {
	Get() []func(*T) error
}
```

Then implement all of the option-bearing structs as "Optioners":

```go
// Optioners is a running list of optional data than can be merged using the
// OptionsGetter.
type Optioners interface {
	FindArgs
}
```

Then we can create a generic merge method:

```go
// Merge will functionally merge a slice of OptionGetters in a "last-one-wins"
// algorithm.
func Merge[T Optioners](opts []Getter[T]) (*T, error) {
	t := new(T)
	for _, opt := range opts {
		for _, setOpt := range opt.Get() {
			if err := setOpt(t); err != nil {
				return t, err
			}
		}
	}

	return t, nil
}
```

Then add the Get method to the options:

```go
// Get will return the Opts list from FindOptions.
func (fo FindOptions) Get() []func(*FindArgs) error {
	return fo.Opts
}
```

And the find method changes to this:

```go
func (coll *Collection) Find(ctx context.Context, filter interface{}, opts ...options.Getter[options.FindArgs]) (cur *Cursor, err error) {

	fo, err := options.Merge[options.FindArgs](opts)
	if err != nil {
		return nil, err
	}
}
```

I've create a POC for this pattern using the latest commit on this PR [here](https://github.com/prestonvasquez/mongo-go-driver/tree/GODRIVER-2696-gen-prop).

Reply:
Additionally, we can seal the Getter by not exporting get and keeping the logic internal.

Reply:
That API proposal looks great! I see @qingyang-hu has integrated it into this PoC. I'm not currently convinced that sealing the `Getter` interface (called `Options` in this PR) is necessary.